data_import: Refactor the uploads processing code paths.

PieterCK opened the pull request:
This is mainly a prep PR to support converting MS Teams attachments. Changes:
- Update all data import tools to use a common dataclass to create upload record data.
- Update all data import tools to use a common method to construct emoji `path_id` and Markdown link for it.
- Refactor how the Slack importer processes upload files. This is a prep work to make the process reusable for Microsoft Teams uploads.
 - Improve response handling for data import file requests.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Manually tested, no regression found:
_(tested again on November 24)_

| Slack | Imported to Zulip |
|--------|--------|
| <img width="1493" height="340" alt="image" src="https://github.com/user-attachments/assets/2106f887-e8d9-45e5-b5f2-805114ead80c" /> |<img width="1086" height="640" alt="image" src="https://github.com/user-attachments/assets/fd3419b9-a20e-45cd-afcc-18401f57a193" /> | 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
This is ready for review.

alexmv commented:
You'll want to stack this on #36395; the just-merged #28617 unfortunately just caused some conflicts here.

alexmv commented:
This is some very helpful refactoring!  I suspect the last couple commits are going to change notably once you rebase.

PieterCK commented:
Manually tested again, this should be ready for review.

alexmv commented on zerver/data_import/slack.py:
```diff
@@ -1025,6 +1030,7 @@ def channel_message_to_zerver_message(
     list[ZerverFieldsT],
     list[ZerverFieldsT],
+    list[UploadFileRequest],
     list[ZerverFieldsT],
```
This is getting sufficiently unwieldily that a dataclass might be helpful.

PieterCK commented on zerver/data_import/slack.py:
```diff
@@ -1736,7 +1762,7 @@ def do_convert_directory(
     create_converted_data_files(emoji_records, output_dir, "/emoji/records.json")
     create_converted_data_files(avatar_records, output_dir, "/avatars/records.json")
-    create_converted_data_files(uploads_records, output_dir, "/uploads/records.json")
+    create_converted_data_files(uploads_list, output_dir, "/uploads/records.json")
```
Added ` import_util: Use get_file_from_url to properly handle file request. ` to fix `get_avatars` and `get_emojis` too.

PieterCK commented on zerver/data_import/import_util.py:
```diff
@@ -659,49 +668,44 @@ def process_avatars(
-    upload_list: list[ZerverFieldsT], output_dir: str, processes: int
-) -> list[ZerverFieldsT]:
+    upload_file_request_list: list[UploadFileRequest], output_dir: str, processes: int
+) -> None:
```
Ahh, totally. I didn't understand `run_parallel_queue` enough to catch your [earlier comment](https://github.com/zulip/zulip/pull/36335#discussion_r2495957103) on this. I think we can also process user avatars while processing users using the same technique. Probably a good follow-up work after this PR

PieterCK replied:
Dropped the previous commit that passes the down a list of `UploadsFileRequest` and used `run_parallel_queue` in this commit:  `slack_importer: Improve uploads processing logic.` 

I've also squashed the commit that adds response handling for file requests to this commit.

alexmv commented on zerver/data_import/slack.py:
```diff
@@ -862,7 +863,7 @@ def convert_slack_workspace_messages(
         total_reactions += reactions
         total_attachments += attachment
-        total_uploads += uploads
+        total_uploads += [asdict(upload) for upload in uploads]
```
I'm curious why we can't propagate this further, without the `asdict`.

Regardless, certainly feels like an improvement.

alexmv commented on zerver/data_import/import_util.py:
```diff
@@ -870,24 +869,15 @@ def validate_user_emails_for_import(user_emails: list[str]) -> None:
 @dataclass
 class AttachmentLinkResult:
-    path: str
+    url: str
```
Why this change?  The term we use pretty consistently is `path_id`, and we're still returning `path_id`, not a URL -- it doesn't start with `/user_uploads/`

alexmv commented on zerver/data_import/slack.py:
```diff
@@ -1691,22 +1708,31 @@ def do_convert_directory(
+    with run_parallel_queue(
+        partial(download_and_export_upload_file, output_dir),
+        processes,
+        initargs=tuple(),
```
This is the default, and should be omitted -- it's irrelevant without an `initializer` anyway.

PieterCK replied:
Omitted `initargs=tuple(),`

alexmv commented on zerver/data_import/import_util.py:
```diff
@@ -863,3 +865,20 @@ def validate_user_emails_for_import(user_emails: list[str]) -> None:
+
+def get_attachment_path_and_content(fileinfo: ZerverFieldsT, realm_id: int) -> tuple[str, str]:
+    # Should be kept in sync with its equivalent in zerver/lib/uploads in the function
+    # 'upload_message_attachment'
```
get_markdown_link_for_url is misplaced (it should be somewhere related to _markdown_, not to the email mirror) but otherwise looks good.

alexmv reviewed (commented):
This is looking good!

Besides a few minor comments, I would also suggest doing a pass on your commit messages, to make sure they think about if they need to address the _why_, for changes which aren't straightforward refactorings.

For instance, the switch to `get_markdown_link_for_url` should mention that it's fixing a class of bugs with `[]` in the filenames.  The commit which switches to `run_parallel_queue` should not just say it "improves upload processing logic" -- the _logic_ (i.e. what gets processed, and why) is entirely unchanged.  The commit message should talk about how it moves the downloading to be in parallel with the message conversion.

One reason I particularly appreciate the process of writing commit messages is to catch my own latent bugs or misunderstandings.  For instance, in thinking about how I would argue for the `upload_backend.generate_message_upload_path` change, I thought about claiming that it would fix upload paths in conversions to be suitable to their backends.  But the act of _thinking_ about that argument made me realize that _that argument is wrong_.  `zerver/lib/import_realm.py` already uses `upload_backend.generate_message_upload_path` during import, to remap the arbitrary import paths to the right place (since, at very least, the realm_id obviously cannot be known at conversion time, and is set to 0).

Which is all to say that I think I convinced myself that `` import_util: Update `get_attachment_path_and_content`. `` (6c8991ce6b7c423f3dcfe1ace87f990dd140f9b4) is incorrect to do -- sorry for leading you astray!  It's not _entirely_ wrong, in that it won't produce wrong results.  But there's no reason that the paths of attachments in the converted archive should be based on whatever upload backend the _converting server_ happened to be configured with -- that's irrelevant.  Using the local-file-storage algorithm (as it was doing previously), which provides fanout to avoid inode exhaustion, seems much more arguably correct.  Though if we leave it like that, we should update the "Should be kept in sync" comment there, to clarify that the file layout here is arbitrary, and just for the written-out import archive, and remove the implication that it needs to be kept in sync with anything.  And we certainly shouldn't use the misleading local variable `s3_path` for something which is not written to S3.

PieterCK commented:
> For instance, the switch to get_markdown_link_for_url should mention that it's fixing a class of bugs with [] in the filenames.

  OK! Updated the commit description to:
  ```
  import_util: Use get_markdown_link_for_url to create attachment link.
  
  This now doesn't generate buggy filenames containing `[`, `]`.
  ```

PieterCK commented:
>  The commit which switches to run_parallel_queue should not just say it "improves upload processing logic" -- the logic (i.e. what gets processed, and why) is entirely unchanged. The commit message should talk about how it moves the downloading to be in parallel with the message conversion.
  
  Updated to:
  ```
  slack_importer: Parallelize Slack uploads data processing.
  
  This moves the uploads downloading process to be in parallel with the
  message conversion.
  ```

PieterCK commented:
Other minor tweaks related to commit descriptions:
- Squashed the commit that moves `build_uploads` to the commit that replaces it with `UploadsRecordData`.
- Renamed the commits that refactor a function to return a dataclass to:
  - `import_util: Return an AttachmentLinkResult instead of a tuple.` 
  - `slack_importer: Return a MessageConversionResult instead of a tuple.`

PieterCK commented:
Resolved merge conflict

alexmv commented on zerver/data_import/mattermost.py:
```diff
@@ -1046,7 +1048,7 @@ def do_convert_data(mattermost_data_dir: str, output_dir: str, masking_content:
         attachment: dict[str, list[Any]] = {"zerver_attachment": zerver_attachment}
-        create_converted_data_files(uploads_list, realm_output_dir, "/uploads/records.json")
         create_converted_data_files(attachment, realm_output_dir, "/attachment.json")
+        create_converted_data_files(uploads_list, realm_output_dir, "/uploads/records.json")
```
Is this reordering intentional?  If so, why?

timabbott replied:
Seems like it can't possibly have a functional effect. But it does place the `attachment` table closer to where it's used, so is probably positive. Will merge with this as-is.
